From: Denys Berkovskyy <berkovskyy@gmail.com>
Date: Fri, 17 Nov 2017 03:29:42 +0000
Subject: Fix inclusion of standard headers on non-linux systems

---
 src/main.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/main.cpp b/src/main.cpp
index ad87adc..91a3afe 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -25,8 +25,10 @@ GNU General Public License for more details.
 #include <string>
 #include <sys/stat.h>
 #include <time.h>
-#ifdef OS_LINUX
+#ifdef HAVE_STDLIB_H
  #include <stdlib.h>
+#endif
+#ifdef HAVE_STRING_H
  #include <string.h>
 #endif
 #include "scsiencrypt.h"
