From: Denys Berkovskyy <berkovskyy@gmail.com>
Date: Sun, 19 Nov 2017 22:33:40 +0000
Subject: Update AC_INIT and AM_INIT_AUTOMAKE invocations in configure.ac

---
 configure.ac | 9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

diff --git a/configure.ac b/configure.ac
index ac9b9d2..99cf1c6 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1,10 +1,10 @@
-AC_INIT(src/main.cpp)
-
 PACKAGE=stenc
 VERSION=1.0.7
 
+AC_INIT($PACKAGE, $VERSION)
+AC_CONFIG_SRCDIR([src/main.cpp])
+AM_INIT_AUTOMAKE
 
-AM_INIT_AUTOMAKE($PACKAGE, $VERSION)
 AC_CONFIG_HEADERS([config.h])
 AC_CHECK_HEADER([sys/types.h])
 AC_CHECK_HEADER([sys/machine.h])
@@ -45,9 +45,6 @@ AC_ARG_WITH([default-key-size],
 	    [AC_DEFINE_UNQUOTED([DEFAULT_KEYSIZE],$withval,"") AC_MSG_RESULT($withval)],
 	    [AC_DEFINE([DEFAULT_KEYSIZE],32,"") AC_MSG_RESULT(32)])
 
-	
-				     
-
 AC_MSG_CHECKING(your OS)
 system=`uname`
 case $system in
